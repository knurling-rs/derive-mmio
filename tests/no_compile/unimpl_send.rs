#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct CpuPrivateRegBlock {
    // this is read-write by default
    data: u32,
}

#[negative_impl::negative_impl]
impl !Send for CpuPrivateRegBlock {}

fn send_check<T: Send>(_: &T){}

fn main() {
    let mut private_peripheral = CpuPrivateRegBlock {
        data: 0xA,
    };

    // Safety: We're pointing at a real object
    let mmio_uart = unsafe { CpuPrivateRegBlock::new_mmio(core::ptr::addr_of_mut!(private_peripheral)) };
    send_check(&mmio_uart);
}
