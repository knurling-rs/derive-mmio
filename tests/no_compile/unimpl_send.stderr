error[E0277]: `*const ()` cannot be sent between threads safely
 --> tests/no_compile/unimpl_send.rs:1:10
  |
1 | #[derive(derive_mmio::Mmio)]
  |          ^^^^^^^^^^^^^^^^^ `*const ()` cannot be sent between threads safely
  |
  = help: the trait `Send` is not implemented for `*const ()`
note: required for `__Wrapper<'_, *const ()>` to implement `Send`
 --> tests/no_compile/unimpl_send.rs:8:1
  |
8 | #[negative_impl::negative_impl]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: 1 redundant requirement hidden
  = note: required for `CpuPrivateRegBlock` to implement `Send`
  = help: see issue #48214
  = note: this error originates in the derive macro `derive_mmio::Mmio` which comes from the expansion of the attribute macro `negative_impl::negative_impl` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `*const ()` cannot be sent between threads safely
  --> tests/no_compile/unimpl_send.rs:20:16
   |
20 |     send_check(&mmio_uart);
   |     ---------- ^^^^^^^^^^ `*const ()` cannot be sent between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Send` is not implemented for `*const ()`
note: required for `__Wrapper<'_, *const ()>` to implement `Send`
  --> tests/no_compile/unimpl_send.rs:8:1
   |
8  | #[negative_impl::negative_impl]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: 2 redundant requirements hidden
   = note: required for `MmioCpuPrivateRegBlock<'_>` to implement `Send`
note: required by a bound in `send_check`
  --> tests/no_compile/unimpl_send.rs:11:18
   |
11 | fn send_check<T: Send>(_: &T){}
   |                  ^^^^ required by this bound in `send_check`
   = note: this error originates in the attribute macro `negative_impl::negative_impl` (in Nightly builds, run with -Z macro-backtrace for more info)
